src/App.tsx

```diff
@@ -10,51 +12,150 @@ export function App() {
   const [activeTabIndex, setActiveTabIndex] = useState(0);
   const [platform, setPlatform] = useState<browser.runtime.PlatformOs>("mac");
   const [search, setSearch] = useState<string>("");
-  const [tabs, setTabs] = useState<browser.tabs.Tab[]>([]);
+  const [tabs, setTabs] = useState<SearchResult[]>([]);
+  const [currentSearchEngine, setCurrentSearchEngine] = useState<
+    string | undefined
+  >(undefined);
 
   const resultsListref = useRef<HTMLDivElement>(null);
+  const searchInputRef = useRef<HTMLInputElement>(null);
 
   const debouncedTerm = useDebounce(search, 150);
   const arrowDownPressed = useKeyPress("ArrowDown");
   const arrowUpPressed = useKeyPress("ArrowUp");
 
-  const getInitialTabs = async () => {
-    const tabs = await browser.tabs.query({});
-    setTabs(tabs);
+  const getAllBookmarks = async (): Promise<
+    browser.bookmarks.BookmarkTreeNode[]
+  > => {
+    const bookmarkTree = await browser.bookmarks.getTree();
+    const bookmarks: browser.bookmarks.BookmarkTreeNode[] = [];
+
+    const traverse = (nodes: browser.bookmarks.BookmarkTreeNode[]) => {
+      for (const node of nodes) {
+        // Only include actual bookmarks (not folders)
+        if (node.url) {
+          bookmarks.push(node);
+        }
+
+        if (node.children) {
+          traverse(node.children);
+        }
+      }
+    };
+
+    traverse(bookmarkTree);
+    return bookmarks;
+  };
+
+  const getFaviconUrl = (url: string | undefined): string | undefined => {
+    if (!url) return undefined;
+
+    try {
+      const urlObj = new URL(url);
+      return `${urlObj.protocol}//${urlObj.hostname}/favicon.ico`;
+    } catch {
+      return undefined;
+    }
   };
 
-  const activateTab = async (tab: browser.tabs.Tab) => {
-    if (!tab.id) {
-      return;
+  const findDefaultEngine = async () => {
+    const engines = await browser.search.get();
+    const defaultEngine = engines.find((e) => e.isDefault);
+    if (defaultEngine) {
+      setCurrentSearchEngine(defaultEngine.name);
     }
+  };
+
+  const getInitialTabs = async () => {
+    const [browserTabs, bookmarks] = await Promise.all([
+      browser.tabs.query({}),
+      getAllBookmarks(),
+    ]);
 
-    await browser.tabs.update(tab.id, { active: true });
+    const tabResults: SearchResult[] = browserTabs.map((tab) => ({
+      type: "tab",
+      favIconUrl: tab.favIconUrl,
+      title: tab.title || "Untitled",
+      url: tab.url,
+      tabId: tab.id,
+    }));
+
+    setTabs(tabResults);
+  };
+
+  const activateSearchResult = async (result: SearchResult) => {
+    if (result.type === "tab" && result.tabId) {
+      await browser.tabs.update(result.tabId, { active: true });
+    } else if (result.type === "bookmark") {
+      await browser.tabs.create({ url: result.url });
+    } else if (result.type === "search") {
+      await browser.search.query({
+        text: result.title,
+        disposition: "NEW_TAB",
+      });
+    }
```

Comment:
```suggestion

    const resultTypes = {
      'tab': (result) => await browser.tabs.update(result.tabId, { active: true }),
      'url': (result) => await browser.tabs.create({ url: result.url }),
      'bookmark': (result) => {},
      'search': (result) => await browser.search.query({ text: result.title, disposition: "NEW_TAB" })
    }

    return resultTypes?.[result.type](result)
```

nit: mostly readability